### PR TITLE
User Agent & Programmatic Configuration

### PR DESCRIPTION
**Pull request by adamzr**
- Add Java version to default User Agent
 - Add the library version to the user agent
 - Add the OTJ version to the user agent
 - Allow User Agent to be customizable via config
 - Allow config to be passed in via code as an alternative to config
based setup (useful for adding config based on runtime)

### CODE REVIEW

File: clientfactory-resteasy/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,24 @@
+Manifest-Version: 1.0
+Created-By: Apache Maven 3.6.0
+Built-By: aricheimer
```
Comment (mikebell90) [inline clientfactory-resteasy/META-INF/MANIFEST.MF]: ?

**Comment by mikebell90** [comment]
- Maven Project good, test with demo
- user agent is inflexible, lots of code, and major security issue. 
- Ability to programmatically - try Immutables

**Comment by mikebell90** [comment]
* All cients need implementation (jaxrs rest, reactive)

File: client/src/main/java/com/opentable/jaxrs/JaxRsClientConfiguration.java [hunk 2]
```diff
@@ -15,8 +15,8 @@
 public class JaxRsClientConfiguration {
 
     @Bean
-    SpecializedConfigFactory<JaxRsClientConfig> jaxRsConfigFactory(PropertyResolver pr) {
-        return SpecializedConfigFactory.create(pr, JaxRsClientConfig.class, "jaxrs.client.${name}");
+    SpecializedConfigFactory<JaxRsClientConfig> jaxRsConfigFactory(ConfigurableEnvironment environment) {
+        return SpecializedConfigFactory.create(environment, JaxRsClientConfig.class, "jaxrs.client.${name}");
```
Comment (adamzr) [inline client/src/main/java/com/opentable/jaxrs/JaxRsClientConfiguration.java]: Unrelated, but fixes the deprecation

File: clientfactory-resteasy/src/main/java/org/jboss/resteasy/client/jaxrs/JettyResteasyClientBuilder.java [hunk 3]
```diff
@@ -42,6 +44,7 @@ public ResteasyClient build() {
         client.setConnectTimeout(config.getConnectTimeout().toMillis());
         client.setMaxConnectionsPerDestination(config.getHttpClientDefaultMaxPerRoute());
         client.setRemoveIdleDestinations(true);
+        client.setUserAgentField(new HttpField(HttpHeader.USER_AGENT, config.getUserAgent()));
```
Comment (mikebell90) [inline clientfactory-resteasy/src/main/java/org/jboss/resteasy/client/jaxrs/JettyResteasyClientBuilder.java]: Log it explicitly.

Comment (adamzr) [reply clientfactory-resteasy/src/main/java/org/jboss/resteasy/client/jaxrs/JettyResteasyClientBuilder.java]: Added logging

File: client/pom.xml [hunk 4]
```diff
@@ -93,6 +93,24 @@
       <artifactId>javax.ws.rs-api</artifactId>
     </dependency>
 
+    <dependency>
+        <groupId>com.google.code.findbugs</groupId>
+        <artifactId>findbugs-annotations</artifactId>
+    </dependency>
+
+    <dependency>
+        <groupId>com.google.errorprone</groupId>
```
Comment (mikebell90) [inline client/pom.xml]: ?

File: clientfactory-resteasy-apache/src/main/java/com/opentable/jaxrs/JaxRsClientFactoryImpl.java [hunk 5]
```diff
@@ -106,12 +106,15 @@ public static HttpClientBuilder prepareHttpClientBuilder(String clientName, JaxR
         connectionManager.setMaxTotal(config.getConnectionPoolSize());
         connectionManager.setDefaultMaxPerRoute(config.getHttpClientDefaultMaxPerRoute());
 
+        LOG.info("Setting User-Agent for the {} HTTP client to {}", clientName, config.getUserAgent());
```
Comment (mikebell90) [inline clientfactory-resteasy-apache/src/main/java/com/opentable/jaxrs/JaxRsClientFactoryImpl.java]: include jetty version ;)

File: client/src/main/java/com/opentable/jaxrs/JaxRsClientConfig.java [hunk 6]
```diff
@@ -171,8 +186,22 @@ default int getProxyPort() {
...
      */
+    @Value.Default
     default boolean isDisableTLS13() {
         return false;
     }
 
+    /**
+     * Get a user agent for the client to use
+     *
+     * Supported: resteasy, resteasy-apache
```
Comment (mikebell90) [inline client/src/main/java/com/opentable/jaxrs/JaxRsClientConfig.java]: I believe you are lying here?

Comment (adamzr) [reply client/src/main/java/com/opentable/jaxrs/JaxRsClientConfig.java]: Why? I added support for both of those.

**Pull request merged**